add initial devlinkd implementation

jpirko opened the pull request:
# Devlinkd: Systemd Service for Devlink Configuration

## Overview

This PR introduces `systemd-devlinkd`, a new systemd service that provides declarative configuration management for Linux devlink objects. Devlink is a kernel interface for configuring advanced networking hardware features found in modern NICs (e.g., NVIDIA, Intel, Broadcom), including E-Switch modes, port splitting, device parameters, and health monitoring.

## Problem Statement

Currently, configuring devlink objects requires manual execution of `devlink` CLI commands, often in custom scripts or initramfs hooks. This approach has several limitations:

- **No persistence**: Configuration is lost on reboot unless manually scripted
- **Race conditions**: Manual scripts may run before devices are ready
- **Inconsistent patterns**: Different users/distributions may implement devlink configuration differently
- **No declarative management**: Configuration is scattered across scripts rather than centralized

This follows up on a discussion from 5 years ago: https://github.com/systemd/systemd/issues/15782

## Solution

`systemd-devlinkd` provides a declarative, systemd-native approach to devlink configuration that:

- **Persists configuration** across reboots via `.devlink` configuration files
- **Integrates with systemd boot process** (runs before `systemd-networkd`, supports initrd transitions)
- **Follows systemd patterns** similar to `systemd-networkd` and others
- **Supports hotplugging** - configures devices as they appear
- **Provides drop-in support** for easy customization without modifying vendor files
- **Match-action** behavior makes sure that only objects referenced by configuration files are touched. The rest is ignored. If there are no configuration files present, `systemd-devlinkd` run has no effect.

## Supported Object Types

This initial implementation supports four devlink object types:

1. **`dev`** - Main devlink device objects (e.g., PCI devices)
   - Configuration: E-Switch mode (legacy/switchdev)

2. **`port`** - Port objects (physical ports, E-Switch VF links)
   - Configuration: Port splitting (e.g., split 100G port into 4x25G)

3. **`param`** - Device/port parameters
   - Configuration: Parameter values (e.g., ACL region rehash interval, keep link up)

4. **`health_reporter`** - Health monitoring reporters
   - Configuration: Grace period, auto-recovery, auto-dump settings

The architecture is extensible - additional object types can be added by implementing the `DevlinkVTable` interface, and match criteria can be extended by implementing the `DevlinkMatchVTable` interface.

## Use Cases and Examples

### 1. E-Switch Configuration (SR-IOV)

Configure a NIC to use switchdev mode for SR-IOV:

```ini
# /etc/systemd/devlink/10-eswitch.devlink
[Match]
Kind=dev
Handle=pci/0000:01:00.0

[Dev]
EswitchMode=switchdev
```

**Benefit**: Enables SR-IOV functionality automatically on boot, without manual intervention.

### 2. Port Splitting

Split a 100G port into 4x25G ports:

```ini
# /etc/systemd/devlink/10-port-split.devlink
[Match]
Kind=port
Handle=pci/0000:01:00.0
Index=1

[Port]
SplitCount=4
```

**Benefit**: Automatically configure port topology when splitter cable is used to connect to multiple hosts/switches.

### 3. Device Parameters

Configure device-specific parameters (e.g., ACL region rehash interval for mlxsw):

```ini
# /etc/systemd/devlink/10-param-rehash.devlink
[Match]
Kind=param
Handle=pci/0000:01:00.0
Name=acl_region_rehash_interval

[Param]
Value=3000
```

**Benefit**: Tune device behavior declaratively, with configuration surviving reboots.

### 4. Health Reporter Configuration

Configure automatic health monitoring and recovery:

```ini
# /etc/systemd/devlink/10-health-fw-fatal.devlink
[Match]
Kind=health_reporter
Handle=pci/0000:01:00.0
Name=fw_fatal

[HealthReporter]
GracePeriod=40000
AutoRecover=false
AutoDump=true
```

**Benefit**: Centralized health monitoring configuration, with future potential for integration with systemd's monitoring infrastructure.

### 5. Network Interface-Based Matching

Match ports by network interface name:

```ini
# /etc/systemd/devlink/10-ifname-port-split.devlink
[Match]
Kind=port
NetdevName=enp0s1np1

[Port]
SplitCount=4
```

**Benefit**: Sometimes more practical camparing to use of PCI addresses.

## Architecture

- **Modular design**: Each object type implements a `DevlinkVTable` interface
- **Event-driven**: Uses netlink generic Netlink channel to monitor devlink object changes
- **Configuration parsing**: Leverages systemd's existing configuration parser infrastructure
- **State management**: Tracks object state and applies configuration as objects appear
- **Graceful shutdown**: Configuration persists when service stops (supports initrd transitions)

## Testing

Tested with:
- **netdevsim** (Linux kernel testing framework)
- **mlx5** (NVIDIA ConnectX/BlueField drivers)
- **mlxsw** (NVIDIA Spectrum switch drivers)

## Future Work

The codebase is designed to support future enhancements:

- **Health reporter monitoring**: Automatic dump collection and integration with systemd's logging/monitoring
- **Additional object types**: Easy to extend with new devlink object kinds as kernel support evolves
- **D-Bus API**: Full D-Bus interface for runtime configuration (foundation already present)

## Documentation

Complete documentation is provided:
- `systemd-devlinkd.service(8)` - Service documentation
- `systemd.devlink(5)` - Configuration file format with examples

## References

- Kernel devlink documentation: https://docs.kernel.org/networking/devlink/
- Devlink man page: https://man7.org/linux/man-pages/man8/devlink.8.html
- Original discussion: https://github.com/systemd/systemd/issues/15782

---

**Please review**: This implementation follows systemd coding patterns and integrates cleanly with the existing codebase. The architecture is extensible and ready for future enhancements. Configuration examples are provided in the man pages.

This addresses a real user need - many organizations currently rely on custom scripts and initramfs hooks to configure devlink devices, which are fragile, hard to maintain, and don't integrate well with systemd's boot process. By providing a native systemd solution, users can replace these custom scripts with declarative configuration files that are more reliable, easier to maintain, and properly integrated with the systemd ecosystem.

I commit to maintaining this codebase after it is merged, including addressing issues, reviewing patches, and extending functionality as kernel support evolves.

Commit:
devlinkd: add devlinkd base

Introduce a skeleton of devlinkd. This prepares the infrastructure in
order to introduce individual matches and objects in follow-up patchset.

There is one configuration file per object. Devlinkd first loads
configuration files. Then it processes incoming devlink netlink messages,
matches against the "Match*" sections in configuration and processes
per object and per message type actions.

Commit:
devlinkd: add devlinkd base

Introduce a skeleton of devlinkd. This prepares the infrastructure in
order to introduce individual matches and objects in follow-up patchset.

There is one configuration file per object. Devlinkd first loads
configuration files. Then it processes incoming devlink netlink messages,
matches against the "Match*" sections in configuration and processes
per object and per message type actions.

Commit:
sd-netlink: introduce variable type

In order to pass by policy for attributes that dynamically change type,
introduce variable type and avoid type checking for it.

Commit:
sd-netlink: introduce nested union by u8 type

Similar to other nested union types, introduce another one, this time
the selection is made by u8 attribute.

Commit:
sd-netlink: introduce netlink_attributes envelope

In preparation for allowing to parse one attribute type multiple times,
introduce netlink_attributes envelope structure.

Commit:
sd-netlink: add cleanup for struct netlink_attributes

In preparation for allowing to parse one attribute type multiple times,
add a cleanup for previously introduced netlink_attributes envelope
structure.

Commit:
sd-netlink: allow to parse multiple attributes with same attr_type

Benefit from previous two preparation patches and instead of single
offset per attr_type store array of offsets. Introduce
get_attributes_count() helper to get the number of attributes.

Commit:
sd-netlink: allow to parse multiple attributes with same attr_type

Benefit from previous two preparation patches and instead of single
offset per attr_type store array of offsets. Introduce
get_attributes_count() helper to get the number of attributes.

Commit:
sd-netlink: introduce indexed variant of string read

Benefit from the previous patch and allow user to read multiple string
values of same attribute by passing index.

Commit:
sd-netlink: introduce indexed variant of string read

Benefit from the previous patch and allow user to read multiple string
values of same attribute by passing index.

Commit:
sd-netlink: introduce indexed variant of enter container

Benefit from previously introduced multi-attribute infrastructure
and allow user to enter multiple containers of same attribute
by passing index.

Commit:
sd-netlink: introduce indexed variant of enter container

Benefit from previously introduced multi-attribute infrastructure
and allow user to enter multiple containers of same attribute
by passing index.

Commit:
sd-netlink: use new indexed function in read_strv()

Instead of doing manual things manually, benefit from previously
introduced multi-attribute infrastructure and make read_strv() simpler.

Commit:
sd-netlink: use new indexed function in read_strv()

Instead of doing manual things manually, benefit from previously
introduced multi-attribute infrastructure and make read_strv() simpler.

Commit:
sd-netlink: introduce read_s32() helper

In order to be able to read signed 32-bit attributes, introduce a helper
for the job.

Commit:
devlinkd: add health reporter object

Introduce health reporter port object, allowing to work with devlink
health reporter. Initially allow to configure grace perion, auto dump
and auto recover options.

Commit:
devlinkd: add health reporter object

Introduce health reporter port object, allowing to work with devlink
health reporter. Initially allow to configure grace perion, auto dump
and auto recover options.

Commit:
devlinkd: add health reporter match

Introduce a devlink health reporter object match, allowing to match
on HEALTH_REPORTER_NAME attribute in devlink netlink messages.

Commit:
devlinkd: add health reporter match

Introduce a devlink health reporter object match, allowing to match
on HEALTH_REPORTER_NAME attribute in devlink netlink messages.

Commit:
devlinkd: add param object

Introduce devlink param object, allowing to set value of devlink param
and devlink port param.

Commit:
devlinkd: add param object

Introduce devlink param object, allowing to set value of devlink param
and devlink port param.

Commit:
devlinkd: add param match

Introduce a devlink param object match, allowing to match on PARAM_NAME
attribute in devlink netlink messages.

Commit:
devlinkd: add param match

Introduce a devlink param object match, allowing to match on PARAM_NAME
attribute in devlink netlink messages.

Commit:
devlinkd: add port object

Introduce devlink port object, allowing to work with devlink port.
Initially allow to configure port split.

Commit:
devlinkd: add port object

Introduce devlink port object, allowing to work with devlink port.
Initially allow to configure port split.

Commit:
devlinkd: add port ifname match with ifname tracker

Introduce an ifname tracker that uses RT netlink and stores netdevice
ifnames. Introduce port ifname match that uses the tracker to get
devlink port attributes by ifname.

Commit:
devlinkd: add port ifname match with ifname tracker

Introduce an ifname tracker that uses RT netlink and stores netdevice
ifnames. Introduce port ifname match that uses the tracker to get
devlink port attributes by ifname.

Commit:
devlinkd: add internal port cache object

In order to be able to support ifname-based matching in a follow-up
patch, introduce an internal port cache object to store devlink port
attributes and allow them to be looked-up by ifindex.

Commit:
devlinkd: add internal port cache object

In order to be able to support ifname-based matching in a follow-up
patch, introduce an internal port cache object to store devlink port
attributes and allow them to be looked-up by ifindex.

Commit:
devlinkd: add port match

Introduce a devlink port object match, allowing to match on PORT_INDEX
and PORT_SPLIT_GROUP attributes in devlink netlink messages.

Commit:
devlinkd: add port match

Introduce a devlink port object match, allowing to match on PORT_INDEX
and PORT_SPLIT_GROUP attributes in devlink netlink messages.

Commit:
devlinkd: add internal reload object

Some of configuration, like set of driverinit param, require devlink
instance reload operation to be performed. To support that, introduce
an internal reload object, allowing to schedule and track
requests of devlink dev reload. Note this processes the same group
of netlink messages as devlink dev object.

Commit:
devlinkd: add internal reload object

Some of configuration, like set of driverinit param, require devlink
instance reload operation to be performed. To support that, introduce
an internal reload object, allowing to schedule and track
requests of devlink dev reload. Note this processes the same group
of netlink messages as devlink dev object.

Commit:
devlinkd: add internal nested object

Introduce in internal nested object, allowing to to track
hierarchy of devlink dev instances. Note this processes the same group
of netlink messages as devlink dev object.

Commit:
devlinkd: add internal nested object

Introduce in internal nested object, allowing to to track
hierarchy of devlink dev instances. Note this processes the same group
of netlink messages as devlink dev object.

Commit:
devlinkd: add dev object

Introduce devlink dev object, allowing to work with devlink instance.
Initially allow to configure eswitch mode.

Commit:
devlinkd: add dev object

Introduce devlink dev object, allowing to work with devlink instance.
Initially allow to configure eswitch mode.

Commit:
devlinkd: add dev match

Introduce a devlink dev object match, allowing to match on BUS_NAME and
DEV_NAME attributes in devlink netlink messages.

Commit:
devlinkd: add dev match

Introduce a devlink dev object match, allowing to match on BUS_NAME and
DEV_NAME attributes in devlink netlink messages.

Commit:
units: add systemd-devlinkd.service

Commit:
units: add systemd-devlinkd.service

Commit:
man: add systemd-devlinkd.service manpage

Commit:
man: add systemd-devlinkd.service manpage

Commit:
man: add systemd.devlink manpage

Commit:
man: add systemd.devlink manpage

Commit:
man: systemd.devlink: add dev object

Commit:
man: systemd.devlink: add dev object

Commit:
devlinkd: add common index parse and match

Multiple devlink objects are using index for identification.
Introduce a common parsing function and match callbacks.

Commit:
devlinkd: add common index parse and match

Multiple devlink objects are using index for identification.
Introduce a common parsing function and match callbacks.

Commit:
devlinkd: add common name match

Multiple devlink objects are using name for identification.
Introduce a common match callbacks.

Commit:
devlinkd: add common name match

Multiple devlink objects are using name for identification.
Introduce a common match callbacks.

Commit:
man: systemd.devlink: add port object

Commit:
man: systemd.devlink: add port object

Commit:
man: systemd.devlink: add param object

Commit:
man: systemd.devlink: add param object

Commit:
man: systemd.devlink: add health reporter object

Commit:
man: systemd.devlink: add health reporter object

jpirko commented on src/devlink/devlink-health-reporter.h:
```diff
@@ -0,0 +1,24 @@
+
+typedef struct DevlinkHealthReporter {
+        Devlink meta;
+        int grace_period_valid:1,
```
fixed.

jpirko commented on src/devlink/devlink-health-reporter.h:
```diff
@@ -0,0 +1,24 @@
+typedef struct DevlinkHealthReporter {
+        Devlink meta;
+        int grace_period_valid:1,
+            auto_recover_valid:1,
```
fixed

jpirko commented on src/devlink/devlink-health-reporter.h:
```diff
@@ -0,0 +1,24 @@
+        Devlink meta;
+        int grace_period_valid:1,
+            auto_recover_valid:1,
+            auto_dump_valid:1;
```
fixed

Commit:
sd-netlink: adjust test-netlink to used indexed string read helper

Multiple attributes of the same type are now put into array. To get
the last one use indexed read helper variant with index 9 as the last
one.

Commit:
sd-netlink: adjust test-netlink to used indexed string read helper

Multiple attributes of the same type are now put into array. To get
the last one use indexed read helper variant with index 9 as the last
one.

jpirko commented on src/devlink/devlink.c:
```diff
@@ -0,0 +1,384 @@
+        (void) vsnprintf(pos, len, format, ap);
+        va_end(ap);
+
+        return log_internal(level, error, file, line, func, "%s", (char *) buf);
```
I'm clueless what's wrong here. I tried to silent this, failed multiple times. Any idea?

yuwata replied:
We usually use macro rather than a function to implement such.
See e.g. device-util.h or log-link.h

jpirko commented:
Looks like the currently failing checks, with exception of the log_internal() issue, are not related to this PR.

Commit:
test: add initial tests for devlinkd

Commit:
test: add initial tests for devlinkd

jpirko commented:
> Looks like the currently failing checks, with exception of the log_internal() issue, are not related to this PR.

Pushed with fixes, enhancements, tests, etc. Looks like failing checks are not related to this PR.

nyabinary commented:
> > Looks like the currently failing checks, with exception of the log_internal() issue, are not related to this PR.
> 
> Pushed with fixes, enhancements, tests, etc. Looks like failing checks are not related to this PR.

Should probably rebase now to fix those failing checks?

Commit:
basic/linux: import devlink.h from recent kernel

Commit:
basic/linux: import devlink.h from recent kernel

Commit:
macro: include missing assert include

bluca commented:
The centos mkosi config needs to pull in the `kernel-modules-internal` package for the netdevsim module

jpirko commented:
> The centos mkosi config needs to pull in the `kernel-modules-internal` package for the netdevsim module

@bluca Unless I'm missing something, kernel-modules-internal is not distributed in any centos official repository :(

Commit:
x

Signed-off-by: Jiri Pirko <jiri@nvidia.com>

jpirko commented:
Rebased, cleared-up all the issues found by CI and Claude. I believe I addressed all concerns previously raised. What I need to do now to get this merged please? @bluca @poettering @yuwata ? Thanks!
